File: cryptography/01-CipherLab/src/main/java/no/kalli/Utility.java
package no.kalli;

import picocli.CommandLine;
import picocli.CommandLine.Option;

import static no.kalli.Matrix.*;


/**
 * @author Kristoffer-Andre Kalliainen
 */
@CommandLine.Command(name = "CipherUltimate", footer = "Copyright(c) 2018",
        description = "Encrypt and decrypt message with provided cipher algorithm as argument")
public class Utility {

    private static final int m = 26;

    @Option(names = {"-c", "--cipher"}, required = true, paramLabel = "affine|hill", description = "Select cipher type (affine or hill)")
    private static String type;

    @Option(names = {"-a", "--keyA"}, paramLabel = "keyA", description = "Affine key A")
    private static int a;

    @Option(names = {"-b", "--keyB"}, paramLabel = "keyB", description = "Affine key B")
    private static int b;

    @Option(names = {"-k", "--keyHill"}, paramLabel = "KEY", description = "Hill key. A four letter key.")
    public static String hill;

    public static byte[] key;

    @Option(names = {"-m", "--message"}, paramLabel = "MESSAGE", description = "Message to encrypt")
    public static String message;

    public static byte[] msg;

    @Option(names = {"-h", "--help"}, usageHelp = true, description = "display a help message")
    private boolean helpRequested = false;

    /**
     * Affine encrypt method
     *
     * @param input Byte array to be encrypted
     * @return A affine encrypted byte array
     */
    public static byte[] affineEncrypt(byte[] input) {

        byte[] tmp = new byte[input.length];

        for (int i = 0; i < input.length; i++)
            tmp[i] = (byte) encryptThis(input[i]);

        return tmp;
    }

    /**
     * Affine decrypt
     *
     * @param input Byte array to be decrypted
     * @return A affine decrypted byte array
     */
    public static byte[] affineDecrypt(byte[] input) {

        byte[] tmp = new byte[input.length];

        for (int i = 0; i < input.length; i++)
            tmp[i] = (byte) decryptThis(input[i]);

        return tmp;
    }

    /**
     * Helper method the encrypt one letter
     *
     * @param i The letter as byte format
     * @return The encrypted letter as int
     */
    private static int encryptThis(int i) {
        return ('A' + (a * (i + 'A') + b) % m);
    }

    /**
     * Helper method the decrypt one letter
     *
     * @param i The letter as byte format
     * @return The decrypted letter as int
     */
    private static int decryptThis(int i) {
        int x = modInverse() * ((i - 'A') - b) % m;
        return x < 0 ? ((x + m) % m) + 'A' : x + 'A';
    }

    /**
     * Modulo inverse of the key a
     *
     * @return
     */
    private static int modInverse() {
        a = a % m;
        for (int x = 1; x < m; x++)
            if ((a * x) % m == 1)
                return x;
        return 1;
    }

    /**
     * Hill encrypt
     *
     * @param input Byte array to be encrypted
     * @return Encrypted byte array
     */
    public static byte[] hillEncrypt(byte[] input) {
        var keyMatrix = createKeyMatrix(key);
        var tuplets = createTuples(input);

        var m_res = multiplyMatrix(keyMatrix, tuplets);
        m_res.modulo(m);

        return m_res.toBytes();
    }

    /**
     * Hill decrypt
     *
     * @param input Byte array to be encrypted
     * @return Decrypted byte array
     */
    public static byte[] hillDecrypt(byte[] input) {
        var keyMatrix = createKeyMatrix(key);
        var key_inv = keyMatrix.inverse2x2();
        key_inv.modulo(m);

        var m_res = createTuples(input);

        var c = multiplyMatrix(key_inv, m_res);
        c.modulo(m);
        c.fixNegatives();

        return c.toBytes();
    }


    /**
     * Encrypt a message based on first commandline argument
     *
     * @param plaintext byte array to encrypt
     * @return encrypted byte array
     */
    public static byte[] encryptMessage(byte[] plaintext) {
        byte[] res = new byte[0];
        switch (Utility.type) {
            case "affine":
                res = Utility.affineEncrypt(plaintext);
                break;
            case "hill":
                res = Utility.hillEncrypt(plaintext);
                break;
        }
        return res;
    }

    /**
     * Decrypt a message based on first commandline argument 1
     *
     * @param ciphertext byte array to decrypt
     * @return decrypted byte array
     */
    public static byte[] decryptMessage(byte[] ciphertext) {
        byte[] res = new byte[0];
        switch (Utility.type) {
            case "affine":
                res = Utility.affineDecrypt(ciphertext);
                break;
            case "hill":
                res = Utility.hillDecrypt(ciphertext);
                break;
        }
        return res;
    }

    /**
     * Commandline arguments checker
     *
     * @param args     Commandline arguments
     * @param isServer Is it a client or server?
     */
    static void checkArguments(String[] args, boolean isServer) {
        if (type.equals("hill"))
            key = hill.getBytes();

        if (!isServer)
            msg = message.getBytes();
    }
}

File: cryptography/01-CipherLab/src/main/java/no/kalli/Utility.kt
/**
 *
 */
package no.kalli

import java.nio.charset.Charset
import java.math.BigInteger


/**
 * @author Kristoffer-Andre Kalliainen
 */

data class Affine(val a: Int, val b: Int, val msg: String, var enc: String) : Cipher {

    override fun encrypt(): String {
        return encryptMessage(msg.toByteArray(Charsets.UTF_8)).toString(Charsets.UTF_8)
    }

    override fun decrypt(): String {
        return decryptMessage(enc.toByteArray(Charsets.UTF_8)).toString(Charsets.UTF_8)
    }

    override fun encryptMessage(plaintext: ByteArray): ByteArray {
        var res = ""
        val len = plaintext.toString(Charsets.UTF_8)
        for (c in len)
//            res += ((a * (c - 'a') + b) % 26 + 'a').toChar()

        println(res)
        enc = res
        return res.toByteArray()
    }

    override fun decryptMessage(ciphertext: ByteArray): ByteArray {
        var res = ""

        val inverse = BigInteger.valueOf(a.toLong()).modInverse(BigInteger.valueOf(26))

        val c = ciphertext.toString(Charsets.UTF_8)

        c.forEach {
            val decoded = inverse.toInt() * (it - 'a' - b + 26)
            res  += (decoded % 26 + 'a'.toInt()).toChar()
        }

        println(res)
        return res.toByteArray(Charsets.UTF_8)
    }

    private fun gcd(a: Int, b: Int): Int {
        return if (b != 0) gcd(b, a % b) else a
    }
}

data class Hill(val a: Int, val b: Int, val msg: String) : Cipher {
    override fun encrypt(): String {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun decrypt(): String {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun encryptMessage(plaintext: ByteArray): ByteArray {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun decryptMessage(ciphertext: ByteArray): ByteArray {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}

interface Cipher : IParent {
    fun encrypt(): String
    fun decrypt(): String
}

@DslMarker
annotation class EncryptDsl

@EncryptDsl
class AffineBuilder(initialA: Int, initialB: Int, initialMsg: String) {
    var a: Int = initialA
    var b: Int = initialB
    var msg: String = initialMsg

    fun build(): Affine {
        return Affine(a, b, msg, "")
    }
}

@EncryptDsl
class HillBuilder(initialA: Int, initialB: Int, initialMsg: String) {
    var a: Int = initialA
    var b: Int = initialB
    var msg: String = initialMsg

    fun build(): Hill {
        return Hill(a, b, msg)
    }
}

@EncryptDsl
fun affine(a: Int = 0, b: Int = 0, msg: String = "", setup: AffineBuilder.() -> Unit = {}): Affine {
    val affineBuilder = AffineBuilder(a, b, msg)
    affineBuilder.setup()
    return affineBuilder.build()
}

@EncryptDsl
fun hill(a: Int = 0, b: Int = 0, msg: String = "", setup: HillBuilder.() -> Unit = {}): Hill {
    val hillBuilder = HillBuilder(a, b, msg)
    hillBuilder.setup()
    return hillBuilder.build()
}

fun main(args: Array<String>) {


    val affine = affine {
        a = 5
        b = 7
        msg = "ILOVEPIZZA"
    }

    affine.encrypt()
    affine.decrypt()


}
